bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java

```diff
@@ -640,14 +741,28 @@ private boolean makeWrite(FileStatus file) throws IOException {
     }
 
     private int readInternal(SegmentHandle handle, byte[] buffer, long offset, int bufferOffset, int length) throws IOException {
+        HDFSSegmentHandle hdfsSegmentHandle = asReadableHandle(handle);
+        IOException lastException = null;
         //There is only one file per segment.
-        FileStatus currentFile = findStatusForSegment(handle.getSegmentName(), true);
-        try (FSDataInputStream stream = this.fileSystem.open(currentFile.getPath())) {
-            stream.readFully(offset, buffer, bufferOffset, length);
-        } catch (EOFException e) {
-            throw new IllegalArgumentException(String.format("Reading at offset (%d) which is beyond the current size of segment.", offset));
+        Path targetPath = getFilePath(hdfsSegmentHandle);
+
+        for (int i = 0; i < MAX_ATTEMPT_COUNT; i++) {
```

Comment:
You only need to do this once (update handle if it fails).

Reply:
Newer change does not make changes to current design. fixed.